service/history/api/describeworkflow/api.go

```diff
@@ -178,13 +179,37 @@ func Invoke(
 				LastIndependentlyAssignedBuildId: ai.GetLastIndependentlyAssignedBuildId(),
 			}
 		}
-		if ai.CancelRequested {
-			p.State = enumspb.PENDING_ACTIVITY_STATE_CANCEL_REQUESTED
-		} else if ai.StartedEventId != common.EmptyEventID {
-			p.State = enumspb.PENDING_ACTIVITY_STATE_STARTED
+
+		p.State = workflow.ActivityState(ai)
+		p.Attempt = ai.Attempt
+		if p.Attempt < 1 {
+			p.Attempt = 1
+		}
+
+		if p.State == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED && now.Before(ai.ScheduledTime.AsTime()) {
+			p.NextAttemptScheduleTime = ai.ScheduledTime
+		}
+
+		p.LastAttemptCompleteTime = ai.LastAttemptCompleteTime
+		if !ai.HasRetryPolicy {
+			p.NextAttemptScheduleTime = nil
 		} else {
-			p.State = enumspb.PENDING_ACTIVITY_STATE_SCHEDULED
+			now := shard.GetTimeSource().Now().UTC()
```

Comment:
Don't think you need to convert to UTC here.

Reply:
You also can use the `now` variable you defined above.

Reply:
all this is gone